### ISSUE TITLE
MultipartUpload: method CompleteMultipartUpload doesn't take into account sent PartETags

### ISSUE THREAD

**Issue by ISDemidoff** [adobe/S3Mock#223]
I'm using 2.1.19 (currently latest) version of S3Mock.

When using multipart upload testing I came to a point that this mock implementation doesn't behave same as real Amazon S3 storage:

For example of such behavior, firstly you need to correctly upload three parts (lets name them 1,2,3) after InitUpload.
Then, when calling CompleteMultipartUpload passing only two of them (for example, 1 and 3), **S3Mock will create an object of all 1,2 and 3 parts**. _Amazon S3 would create that object using 1st and 3rd parts as passed to client._

Example JUnit4 test to show incorrect behavior:

```java
import com.adobe.testing.s3mock.junit4.S3MockRule;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.CompleteMultipartUploadRequest;
import com.amazonaws.services.s3.model.InitiateMultipartUploadRequest;
import com.amazonaws.services.s3.model.PartETag;
import com.amazonaws.services.s3.model.S3Object;
import com.amazonaws.services.s3.model.UploadPartRequest;
import org.junit.Assert;
import org.junit.ClassRule;
import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class AdobeS3MockWrongBehavior {
    private static final Random rng = new Random();
    private static final String BUCKET_NAME = "test";
    private static final int BUFFER_SIZE = 128 * 1024;

    // Using convenient junit4 S3Mock Rule to create in-memory instance
    @ClassRule
    public static S3MockRule s3MockRule = S3MockRule.builder()
                                                    .withHttpPort(0)
                                                    .withInitialBuckets(BUCKET_NAME)
                                                    .build();

    // Some helper methods

    private byte[] createRandomBytes() {
        int size = 5 * 1024 * 1024 + rng.nextInt(1024 * 1024);
        byte[] bytes = new byte[size];
        rng.nextBytes(bytes);
        return bytes;
    }

    private UploadPartRequest getUploadPartRequest(String key, String uploadId) {
        return new UploadPartRequest()
                .withBucketName(BUCKET_NAME)
                .withKey(key)
                .withUploadId(uploadId);
    }

    private byte[] readStreamIntoByteArray(InputStream inputStream) throws IOException {
        try (InputStream in = inputStream) {
            ByteArrayOutputStream baos = new ByteArrayOutputStream(BUFFER_SIZE);

            byte[] buffer = new byte[BUFFER_SIZE];
            int bytesRead = -1;
            while ((bytesRead = in.read(buffer)) != -1) {
                baos.write(buffer, 0, bytesRead);
            }
            baos.flush();

            return baos.toByteArray();
        }
    }

    private byte[] concatByteArrays(byte[] arr1, byte[] arr2) {
        byte[] result = new byte[arr1.length + arr2.length];

        System.arraycopy(arr1, 0, result, 0, arr1.length);
        System.arraycopy(arr2, 0, result, arr1.length, arr2.length);

        return result;
    }

    // Actual test, which would be passed green on real Amazon S3 instance

    @Test
    public void wrongBehavior() throws IOException {
        AmazonS3 s3Client = s3MockRule.createS3Client();

        String key = "folder/file.name";

        // Initiate upload
        String uploadId = s3Client.initiateMultipartUpload(new InitiateMultipartUploadRequest(BUCKET_NAME, key))
                                  .getUploadId();

        // Upload 3 parts
        byte[] randomBytes1 = createRandomBytes();
        PartETag partETag1 = s3Client
                .uploadPart(getUploadPartRequest(key, uploadId)
                        .withPartNumber(1)
                        .withPartSize(randomBytes1.length)
                        .withInputStream(new ByteArrayInputStream(randomBytes1)))
                .getPartETag();

        byte[] randomBytes2 = createRandomBytes();
        PartETag partETag2 = s3Client
                .uploadPart(getUploadPartRequest(key, uploadId)
                        .withPartNumber(2)
                        .withPartSize(randomBytes2.length)
                        .withInputStream(new ByteArrayInputStream(randomBytes2)))
                .getPartETag();

        byte[] randomBytes3 = createRandomBytes();
        PartETag partETag3 = s3Client
                .uploadPart(getUploadPartRequest(key, uploadId)
                        .withPartNumber(3)
                        .withPartSize(randomBytes3.length)
                        .withInputStream(new ByteArrayInputStream(randomBytes3)))
                .getPartETag();

        // Adding to parts list only 1st and 3rd part
        List<PartETag> parts = new ArrayList<>();
        parts.add(partETag1);
        parts.add(partETag3);

        // Try to complete with these parts
        s3Client.completeMultipartUpload(new CompleteMultipartUploadRequest(BUCKET_NAME, key, uploadId, parts));

        // Verify only 1st and 3rd counts
        S3Object object = s3Client.getObject(BUCKET_NAME, key);

        // Try to verify content size
        Assert.assertEquals(
                "Content length doesn't match",
                randomBytes1.length + randomBytes3.length,
                object.getObjectMetadata().getContentLength()
        );
        // Try to verify contents
        Assert.assertArrayEquals(
                "Object contents doesn't match",
                concatByteArrays(randomBytes1, randomBytes3),
                readStreamIntoByteArray(object.getObjectContent())
        );
    }
}
```

**Comment by ChaithanyaGK** [comment]
@ISDemidoff Thanks for reporting. I'm on it. I just need to verify few things before raising a PR. Could you please share etag's returned from S3 for uploaded parts and for complete upload. (i.e., `partETag1`, `partETag2`, `partETag3` and `CompleteMultipartUploadResult.eTag`)

**Comment by ISDemidoff** [comment]
@ChaithanyaGK ok, here we go.
I just ran my provided code and got:

uploadId: "9b233d93-87e6-4bfb-bd56-f762ec3100c3"
eTag1: "5973ae733b102b22143dafe3996814ba"
eTag2: "123a2d6c816f1a941f261d02747cdde6"
eTag3: "46f9536e629e5cf77d46358b1aa9b2f7"
completeMultipartUploadResult.getETag: "c250c98b36641df01850f0534afd3d25-3"

Every time I run that snippet, I get different results, but "-3" remains, which I think indicates that 3 parts applied as a file

**Comment by mibollma** [comment]
Seems to introduce some new issue. Since **2.1.21** i'm getting

>  An error occurred (InvalidPart) when calling the CompleteMultipartUpload operation: One or more of the specified parts could not be found. The part might not have been uploaded, or the specified entity tag might not have matched the part's entity tag.

when using **aws s2 sync**. **2.1.20** works fine for me.

**Comment by ChaithanyaGK** [comment]
@mibollma With v2.1.21 we are validating the CompleteMultipartUpload request to be inline with AWS S3 behavior
1. Parts mentioned in the request are previously uploaded
2. If yes, then validating the e-tag of the parts mentioned.

Let me know if you're specifying the proper Multipart parts in the request payload and still facing this issue.

Ref: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html

**Comment by mibollma** [comment]
@ChaithanyaGK I'm not doing anything manual. Just calling **aws s3 sync** on the commandline using aws cli to synchronize a local folder to s3mock, basically populating a bucket with some files that are used during testing. Behaviour is the same with aws cli v1 and v2.

The exact command is
`aws s3 sync <local-folder> s3://<bucket>/<name> --endpoint-url http://localhost:9090`

**Comment by ChaithanyaGK** [comment]
aws-cli is sending e-tag wrapped in quotes, but in our implementation e-tag's are not wrapped which is resulting in this issue. As a workaround I've stripped quotes from e-tag before validation and released v2.1.22

@mibollma @ISDemidoff @mserranom Could you please confirm that your issues are fixed with v2.1.22

**Comment by mibollma** [comment]
@ChaithanyaGK Version **v2.1.22** works as expected again. Thanks alot :thumbsup:

**Comment by mserranom** [comment]
@ChaithanyaGK the new version looks good thanks!

**Comment by ChaithanyaGK** [comment]
Thanks @mibollma & @mserranom for confirming. 
Closing this issue as it is fixed in v2.1.22
